src/client.rs

```diff
@@ -235,7 +243,8 @@ impl RequestBuilder {
         loop {
             let res = {
                 debug!("request {:?} \"{}\"", method, url);
-                let mut req = client.hyper.request(method.clone(), url.clone())
+                let c = client.hyper.lock().unwrap();
```

Comment:
So Mutexes are tricky. This code will surprisingly hold on to the locked client while the request is sent over the network, and the response headers have been received. I'd forgotten that a mutex would play badly here.

This is one of the very few times when a `RwLock` is probably a good idea. The reason is that we have to lock it somehow for when we mutate the `Client`, in the `timeout()` method. But besides that, we actually **don't** want exclusive access, and only require read (`&client`) access. Making use a `RwLock` means that making a change to the timeout has to wait until any current requests are done, but the requests don't block each other.